Add clipboard image data get/set methods

radzo73 opened the pull request:
Closes godotengine/godot-proposals#2949.
TODO:
- [ ] `Image clipboard_get_image`
- [ ] `void clipboard_set_image (Image img)`

spacechase0 committed:
Propogate previously unused NOTIFICATION_WORLD_2D_CHANGED, make CanvasItem/CollisionObject2D use it

Jayman2000 committed:
Streamline and centralize bug reporting guides

Before this change, there was three different bug reporting guides:

- [“Filing an issue on GitHub”][1]
- [“Reporting bugs”][2]
- [The issue template][3]

This commit:

1. makes sure that [the issue template][3] contains all of the same
   information that [“Filing an issue on GitHub”][1] and
   [“Reporting bugs”][2] did and
2. makes [“Reporting bugs”][2] simply tell users to fill out the
   template.

The goal of this change is to make reporting bugs easier. This change
accomplishes that goal by presenting bug reporters with all of the
information they need to know on the bug reporting page itself.

This commit partially implements this proposal:
<https://github.com/godotengine/godot-proposals/discussions/4083>

[1]: https://docs.godotengine.org/en/stable/community/contributing/ways_to_contribute.html#filing-an-issue-on-github
[2]: ./CONTRIBUTING.md#reporting-bugs
[3]: ./.github/ISSUE_TEMPLATE/bug_report.yml

Sauermann committed:
Move call of `push_unhandled_input` from `Window` to `Viewport`

This solves the problem, that mouse events get sent to SubViewports
even if they are outside of the visible area of the SubViewport.

This changes makes SubViewportContainer::unhandled_input redundand.
Shortcut Events now need to be distributed via push_input, in order for
them to be able to reach SubViewports.

kleonc committed:
Make `LocalVector` respect its `tight` template parameter

Rindbee committed:
Improve the UX of ViewportTexture in the editor

The associated `ViewportTexture`s will update the `viewport_path`
in time when the `Viewport`'s nodepath is changed (caused by renaming
the node names or moving in the SceneTree dock).

If the target `Viewport` is changed by resetting the `viewport_path`,
the `ViewportTexture`s will be re-setup and emit `changed` signal in
time.

KoBeWi committed:
Add metadata to resource previews

MewPurPur committed:
Light3D show scaling warning immediately

trollodel committed:
Add multi window code and shader editors

KoBeWi committed:
Add info label to TileMap editor

Wiwip committed:
Modifies the file system dock to use the inline editor instead of a dialog.

*Bugsquad edit:* Closes https://github.com/godotengine/godot-proposals/issues/4933

snibk committed:
Add icon_modulate functionality to PopupMenu

KoBeWi committed:
Add proper snapping to tile polygon editor

Maran23 committed:
Expose horizontal/vertical custom_step as editor property for the ScrollContainer

Chaosus committed:
Add handling of custom visual shader nodes from GDExtension

smix8 committed:
Rework Navigation Avoidance

Rework Navigation Avoidance.

MewPurPur committed:
Use String.repeat() in more places

MewPurPur committed:
Fix popup position of color picker

uuv-rov committed:
Move ShaderWarning definitions into rendering_server

timothyqiu committed:
Allow EditorInspector to change its property name style when necessary

Previously, an EditorInspector's property name can only be set from
outside. Inspectors used for settings needs to respond to changes in
editor settings. So a few boilerplate code is almost always needed,
including watching for a certain editor setting in `_notification()`.

This commit adds a `set_use_settings_style()` function to tell the
inspector to watch for editor settings changes on its own.

yvie-k committed:
Implement vertical icon alignment for buttons

RandomShaper committed:
Avoid interaction issues between resource loading threads

BastiaanOlij committed:
For GDExternal use, provides access to internal graphics handles for textures

dalexeev committed:
Fix `draw_multiline_colors()` for `width < 0`

BastiaanOlij committed:
Fix incorrect HTC action map entries

ChibiDenDen committed:
Fix gles3 rendering on android studio emulator

Use a temporary framebuffer for screen copy from rendertarget to screen.

This solves GLES3 rendering in android studio emulator (before this change there is just a black screen)

Based on discussion in:
https://github.com/godotengine/godot/issues/74828

JohanAR committed:
Add SPECULAR_AMOUNT spatial light shader built-in

Light3D has a light_specular property which is used to set the
intensity of specular contributed by this light source, but it was
previously only used by the default material light shader, and not
possible to use in a custom light() shader.

Chaosus committed:
Fix shader uniform groups to prevent it sorting in alphabet order

Calinou committed:
Improve RenderingServer, RenderingDevice, ShaderGlobalsOverride documentation

This brings the overall class reference completion percentage from 87% to 92%.

SilicDev committed:
Reimplement String.erase

BastiaanOlij committed:
Expose viewports render targer RID

smix8 committed:
Prevent unnecessary navigation map synchronizations

Prevents unnecessary navigation map synchronizations triggered by redundant calls to setters of e.g. region, link or map properties.

m4gr3d committed:
Adds a `scale_gizmo_handles` entry to the `Touchscreen` editor settings

When enabled, this scales the editor icons to improve usability on touchscreen devices.
In addition this commit fixes touch detection for the collision_shape_2d_editor_plugin so it scales with the icons size.

ScorpionInc committed:
Exposes String functions for X509Certificates

Exposes String functions for X509Certificates via two function calls: save_to_string() and load_from_string(str).

KoBeWi committed:
Prevent errors when using ViewportTexture

KoBeWi committed:
Close built-in shaders when closing scene

reduz committed:
Refactor Node Processing

* Node processing works on the concept of process groups.
* A node group can be inherited, run on main thread, or a sub-thread.
* Groups can be ordered.
* Process priority is now present for physics.

This is the first steps towards implementing https://github.com/godotengine/godot-proposals/issues/6424.
No threading or thread guards exist yet in most of the scene code other than Node. That will have to be added later.

RandomShaper committed:
Allow creation of rendering buffers at any time

Faless committed:
mbedTLS: disable weak crypto and TLS versions.

This commit adds a new mbedTLS configuration header to customize the
built-in library (and can be optionally replaced by a platform-specific
one).

Currently, it disables most weak cryptographic functions (with the
notable exceptions of MD5 and SHA-1), along with removing support for
TLS versions 1.0 and 1.1 (making TLSv1.2 the only supported one).

mashumafi committed:
Wait for navigation baking to finish before destruction

rburing committed:
PhysicsDirectBodyState3D: report contact points using global coordinates

Also update the documentation to reflect this in both 2D and 3D.

ChibiDenDen committed:
Add shader cache to gles3

paddy-exe committed:
Fix Rotation Issue with NODE_POSITION_VIEW Visual Shader Node

MewPurPur committed:
Add an example for how to use AudioStreamGenerator

Black-Cat committed:
Fix precision in physics supports generation

Lower threshold for dot was (1.0 - threshold) which is incorrect.
Patch changes it to correct version sqrt(1.0 - threshold * threshold)

Co-authored-by: Ricardo Buring <ricardo.buring@gmail.com>

RandomShaper committed:
Allow concurrent buffering and dispatch of input events

reduz committed:
Add ValidatedCall to MethodBind

* This should optimize GDScript function calling _enormously_.
* It also should simplify the GDScript VM considerably.

NOTE: GDExtension calling performance has most likely been affected until going via ptrcall is fixed.

bonjorno7 committed:
Add API for HSL conversion

Math ported pretty much 1:1 from https://en.wikipedia.org/wiki/HSL_and_HSV
Style doesn't match the existing HSV code exactly, but should be close enough.

dsnopek committed:
Dump API docs from inline GDScript comments using --doctool --gdscript-docs PATH

dalexeev committed:
Fix multiline items drawing in `Tree`

AThousandShips committed:
Fix size error in `BitMap.opaque_to_polygons`

Previous estimate of upper limit on size was incorrect

shendo committed:
[Android] Fix dynamic Variant params stack constructions in JNI callbacks

Emitting signals with params from Android plugins could crash due to
object assignment with uninitialised mem.  Instead, use 'memnew_placement'
to construct into stack addresses.  Make similar JNI callbacks consistent.

Fixes #75754.

RandomShaper committed:
Fix additional cases of breakage of rendering effects

reduz committed:
Redo how the remote filesystem works

Instead of reading files over the network, the new version uses a local file cache and only updates files when it changes.

The original remote filesystem was created 14 years ago, when ethernet was faster than hard drives or even flash. Also, mobile devices have a very small amount of storage.
Nowadays, this is no longer the case so the approach is changed to using a persistent cache in the target device.

Co-authored-by: m4gr3d

MewPurPur committed:
Overhaul the top sections of the class reference (Animation classes)

Tetane committed:
Use a SubViewport for CanvasItem inspector preview

clayjohn committed:
Fix compatibility breakage from adding NoiseTexture3D

Also optimize some of the Noise methods

KoBeWi committed:
Improve and document PackedDataContainer

ChibiDenDen committed:
Add ARB_get_program_binary to glad to support shader cache over desktop GL

stmSi committed:
Added note to used set_deferred while setting scroll values in _ready function.

spanzeri committed:
Fix skeleton_3d & physical_bone_3d editor errors

Tefatika committed:
Fixed error messages when setting all_tab_in_front of TabContainer

Some signal handlers weren't unregistered when removing the wrapped
child TabBar, which resulted in errors printed when the TabBar was
removed and added again when the flag was updated

Calinou committed:
Document high performance cost of turbulence in ParticleProcessMaterial

- Mention that GPUParticlesAttractorVectorField3D can be used as an
  alternative to turbulence in 3D.

radzo73 committed:
Clipboard has/get image methods

A rebase of https://github.com/godotengine/godot/pull/63826

radzo73 committed:
Things I didn't catch before test compiling

boy am I glad I test things

MewPurPur committed:
Add documentation for TubeTrailMesh and RibbonTrailMesh

aaronfranke commented:
CC @Riteo if I recall correctly the clipboard handling needs different code on Wayland so this will affect your port.

HolonProduction committed:
Preserve scene unique names when saving branch as scene.

jmb462 committed:
Fix right click in selection of additional caret

myaaaaaaaaa committed:
Print CPU and GPU usage statistics after Movie Maker is finished

KoBeWi commented:
I now noticed this renames `clipboard_get/set` to `clipboard_get/set_text`. While it's a nice rename, it's way too late for that and the PR won't be accepted with compatibility breakage.

JoNax97 commented:
Could both be kept and the old ones be marked as deprecated? Is that possible for minor releases?

KoBeWi commented:
Yes.
You can mark a method as deprecated by manually adding `is_deprecated="true"` attribute in the XML file.

radzo73 committed:
Methods for iOS

perhaps they work?

radzo73 commented:
@KoBeWi would I also have to re-introduce the old `clipboard_get/set` methods into the code itself?

KoBeWi commented:
Yep, the documentation just mirrors whatever is bound. I think you can bind the same C++ method multiple times with different names.

bruvzg commented on platform/macos/display_server_macos.mm:
```diff
@@ -2101,6 +2102,37 @@
+	NSBitmapImageRep *bitmap = [NSBitmapImageRep imageRepWithData:data];
+	NSData *pngData = [bitmap representationUsingType:NSPNGFileType properties:@{}];
+	image.instantiate();
+	UInt8 buf[pngData.length];
+	[pngData getBytes:buf length:pngData.length];
+	PNGDriverCommon::png_to_image(buf, pngData.length, false, image);
```
```suggestion
	PNGDriverCommon::png_to_image((const uint8_t *)pngData.bytes, pngData.length, false, image);
```

bruvzg commented on platform/ios/display_server_ios.mm:
```diff
@@ -650,16 +652,39 @@ _FORCE_INLINE_ int _convert_utf32_offset_to_utf16(const String &p_existing_text,
+	UIImage *data = pasteboard.image;
+	NSData *pngData = UIImagePNGRepresentation(data);
+	image.instantiate();
+	UInt8 buf[pngData.length];
+	[pngData getBytes:buf length:pngData.length];
+	PNGDriverCommon::png_to_image(buf, pngData.length, false, image);
```
```suggestion
	PNGDriverCommon::png_to_image((const uint8_t *)pngData.bytes, pngData.length, false, image);
```

bruvzg commented on platform/windows/display_server_windows.cpp:
```diff
@@ -341,6 +342,67 @@ String DisplayServerWindows::clipboard_get() const {
+			if (ptr != NULL) {
+				BITMAPINFOHEADER *info = &ptr->bmiHeader;
+				PackedByteArray pba;
+
```
Consider using `resize` and `write` instead of `append` since it can be slow due to multiple reallocation.
```suggestion
	pba.resize(info->biHeight * info->biWidth * 4);
```
and 
```
	pba.write[i] = rgbquad->rgbRed;
	pba.write[i + 1] = rgbquad->rgbGreen;
	pba.write[i + 2] = rgbquad->rgbBlue;
	pba.write[i + 3] = rgbquad->rgbReserved;
```

radzo73 committed:
Update platform/macos/display_server_macos.mm

I trust that at bruvzg knows better than me

Co-authored-by: bruvzg <7645683+bruvzg@users.noreply.github.com>

radzo73 committed:
Update platform/ios/display_server_ios.mm

Co-authored-by: bruvzg <7645683+bruvzg@users.noreply.github.com>

radzo73 committed:
Update platform/windows/display_server_windows.cpp

Co-authored-by: bruvzg <7645683+bruvzg@users.noreply.github.com>

radzo73 committed:
Update display_server_windows.cpp Part 2

Gee golly I hope this works

KoBeWi committed:
Some clarifications on screen-space coordinates

Koyper committed:
Update @GDScript.xml range docs to iterate backwards.

Commit:
Merge pull request #76640 from shendo/android_plugin_crash_on_emit_signal

[Android] Fix dynamic Variant params stack constructions in JNI callbacks

Paulb23 committed:
Move convert_indent into CodeEdit

brno32 committed:
add venv and .venv to the .gitignore

Xenoparrot committed:
Fixed various typos across the docs

smix8 committed:
Fix SurfaceTool::create_from_blend_shape()

Fixes SurfaceTool::create_from_blend_shape().

rburing committed:
Add debug collision shape to CSG with collision

clayjohn committed:
Merge pull request #76597 from Tefatika/fix/tefatika/76578-tab-container-all_tab_in_front-error

Fixed error messages when setting all_tab_in_front of TabContainer

clayjohn committed:
Merge pull request #76598 from Calinou/doc-particleprocessmaterial-turbulence-cost

Document high performance cost of turbulence in ParticleProcessMaterial

brno32 committed:
adjusting step size for fastnoise frequency hint

clayjohn committed:
Merge pull request #68033 from MewPurPur/auto-light3d-warning-updates

Show Light3D's scale warning immediately when a user edits the property.

clayjohn committed:
Use proper UV in cubemap downsampler raster

This removes bias in cubemap downsampling shader that resulted in the bottom of cubemaps being over represented

QbieShay committed:
Merge pull request #76683 from brno32/fastnoise-lite-frequency-hint

adjusting step size for fastnoise frequency hint

MewPurPur committed:
Document TextServerDummy

dsnopek committed:
Explicitly mark inherited export mode when making a dedicated server export

clayjohn committed:
Merge pull request #76609 from HolonProduction/save-branch-as-scene-unique-name

Preserve scene unique names when saving branch as scene.

mhilbrunner committed:
Merge pull request #76555 from Tetane/master

Fix 2D shader preview draws over uniform

mhilbrunner committed:
Merge pull request #76669 from smix8/fix_create_from_blend_shape_4.x

Fix SurfaceTool::create_from_blend_shape()

mhilbrunner committed:
Merge pull request #76692 from clayjohn/Mobile-sky-dark

Use proper UV in cubemap downsampler raster (Fixes reflections in mobile renderer)

mhilbrunner committed:
Merge pull request #76557 from clayjohn/NoiseTexture3D

Fix compatibility breakage from adding NoiseTexture3D

mhilbrunner committed:
Merge pull request #76550 from RandomShaper/vk_gi_take_2

Fix voxel GI issues (2)

mhilbrunner committed:
Merge pull request #76646 from KoBeWi/acoording_to_screen

Some clarifications on screen-space coordinates

RedworkDE committed:
Prevent infinite loop when signal disconnection fails during object deletion.

aaronfranke committed:
Fix small mistake in docs, "antecedents" -> "ancestors"

Commit:
Make sun energy adjustable as a float in the 3D editor's preview sun

AThousandShips committed:
Make documentation sorting use natural order

AThousandShips committed:
Add `naturalcasecmp_to` function to `String`

Functions as a complement to `naturalnocasecmp_to`

clayjohn committed:
Merge pull request #76722 from HotHead007/master

Make sun energy adjustable as a float in the 3D editor's preview sun

spanzeri committed:
Support long path in file access on windows

Changed windows file access file to check for path length and use the \\?\ long format when needed

Commit:
Make environment energy adjustable as a float in 3D editor's preview environment

mhilbrunner committed:
Merge pull request #58518 from Jayman2000/bug-reporting-guidelines

Streamline and centralize bug reporting guides

RandomShaper committed:
Mend some checks in CallQueue

mhilbrunner committed:
Merge pull request #76739 from spanzeri/fix-long-path-windows

mhilbrunner committed:
Merge pull request #76004 from mashumafi/wait-nav-region-thread

mhilbrunner committed:
Merge pull request #75055 from JohanAR/light_specular

mhilbrunner committed:
Merge pull request #76604 from MewPurPur/sylveon-moment-UwU

mhilbrunner committed:
Merge pull request #74883 from dalexeev/fix-thin-multiline-colors

mhilbrunner committed:
Merge pull request #75517 from BastiaanOlij/expose_render_target

mhilbrunner committed:
Merge pull request #76109 from paddy-exe/fix-node-position-view

mhilbrunner committed:
Merge pull request #75864 from KoBeWi/assassin_of_shaders

KoBeWi committed:
Cache feature list in OS.has_feature()

mhilbrunner committed:
Merge pull request #76650 from Koyper/gdscript_range_docs

clayjohn committed:
Merge pull request #75427 from Chaosus/shader_fix_groups_ordering

Fix shader uniform groups to prevent it sorting in alphabet order

clayjohn committed:
Merge pull request #76700 from dsnopek/dedicated-server-mark-inherited

Explicitly mark inherited export mode when making a dedicated server export

clayjohn committed:
Merge pull request #72288 from MewPurPur/use-string-repeat

Use `String.repeat()` to optimize several String methods

clayjohn committed:
Merge pull request #75510 from SilicDev/string_erase

Reimplement `String.erase()` as immutable method

RandomShaper committed:
Avoid redundant locking in CallQueue

RandomShaper committed:
Add missing thread safety to PagedAllocator

novalis committed:
Explain why an image would be invalid for a texture

KoBeWi committed:
Remove erroneous signal connection

clayjohn committed:
Merge pull request #76765 from novalis/explain-why-invalid

Explain why an image would be invalid for a texture

mhilbrunner committed:
Merge pull request #76743 from HotHead007/master

mhilbrunner committed:
Merge pull request #76761 from KoBeWi/this_file_ain't_big_enough_for_both_of_us

mhilbrunner committed:
Merge pull request #76716 from brno32/venv-gitignore

mhilbrunner committed:
Merge pull request #76720 from aaronfranke/ancestortecedent

mhilbrunner committed:
Merge pull request #76699 from MewPurPur/document-text-server-dummy

mhilbrunner committed:
Merge pull request #76664 from Xenoparrot/simple-typos

mhilbrunner committed:
Merge pull request #76373 from MewPurPur/add-AudioStreamGenerator-example

achernik committed:
LSP: don't send empty completion command

OmarShehata committed:
Fix double tap & drag on Android

brno32 committed:
add unit tests for Curve3D

KoBeWi committed:
Fix another collision shape editor crash

funkkiy committed:
Fix building the Windows x86_32 target

In x86_32, `ULONG_PTR` and `SIZE_T` are defined as `unsigned long`, which is not a type supported by Variant without the `NEED_LONG_INT` define.

YeldhamDev committed:
Add more uses of appropriate cursors when resizing/moving some UI nodes

KoBeWi committed:
Don't refresh inspector when changing internal meta

mhilbrunner committed:
Merge pull request #76008 from Calinou/doc-renderingserver-renderingdevice-2

Commit:
Merge pull request #75718 from m4gr3d/add_scale_editor_icons_main

Adds a `scale_gizmo_handles` entry to the `Touchscreen` editor settings

YeldhamDev committed:
Improve some descriptions in the OS docs

Calinou committed:
Document caveats related to exported project binary resource conversion

ajreckof committed:
fix a test for wrapf with signed 32-bit float

AThousandShips committed:
Minor bugfixes to `ItemList`

* Removed unused position code in `gui_input`
* Ensured `set_fixed_icon_size` updates cached size

akien-mga committed:
CI: Make codespell checks blocking, but only check changed files

RandomShaper committed:
Save cluster render shader from being optimized out entirely

akien-mga committed:
Merge pull request #68800 from KoBeWi/tinfo

Add info label to TileMap editor

akien-mga committed:
Merge pull request #70911 from Chaosus/vs_custom_gdextension_support

Add handling of custom visual shader nodes from GDExtension

akien-mga committed:
Merge pull request #72376 from MewPurPur/fix-color-button-popup

Fix popup position of color picker

akien-mga committed:
Merge pull request #74369 from Dragoncraft89/master

Implement vertical icon alignment for buttons

akien-mga committed:
Merge pull request #74930 from BastiaanOlij/fix_action_map_entries

Fix incorrect HTC action map entries

akien-mga committed:
Merge pull request #76399 from RandomShaper/fix_android_input_anr

Allow concurrent buffering and dispatch of input events

akien-mga committed:
Merge pull request #76472 from jmb462/multicarets_selection_popup

Fix right click in selection of additional caret

akien-mga committed:
Merge pull request #76549 from MewPurPur/improve-docs-top-sections-ANIM

Overhaul the top sections of the class reference (Animation classes)

akien-mga committed:
Merge pull request #76561 from KoBeWi/most_obscure_class_that_has_ever_existed

Improve and document PackedDataContainer

akien-mga committed:
Merge pull request #76658 from Paulb23/convert-indent-code-edit

Move convert_indent into CodeEdit

akien-mga committed:
Merge pull request #76728 from RedworkDE/signal-disconnect-loop

Prevent infinite loop when signal disconnection fails during object deletion

akien-mga committed:
Merge pull request #76747 from RandomShaper/fix_mq_redundant

Mend some checks in CallQueue

akien-mga committed:
Merge pull request #76748 from KoBeWi/has_feature(fast)

Cache feature list in `OS.has_feature()`

akien-mga committed:
Merge pull request #76754 from RandomShaper/mq_less_lock

Avoid redundant locking in CallQueue

akien-mga committed:
Merge pull request #76798 from KoBeWi/race_condition_except_there_is_no_thread

Fix another collision shape editor crash

akien-mga committed:
Merge pull request #76808 from davicr/x86-32_build_fix

Fix building the Windows x86_32 target

akien-mga committed:
Merge pull request #76809 from YeldhamDev/cursor_mania

Add more uses of appropriate cursors when resizing/moving some UI nodes

akien-mga committed:
Merge pull request #76822 from YeldhamDev/os_docs

Improve some descriptions in the OS docs

akien-mga committed:
Merge pull request #76828 from akien-mga/ci-codespell-blocking

CI: Make codespell checks blocking, but only check changed files

akien-mga committed:
Merge pull request #70488 from KoBeWi/SNAP!

Add proper snapping to tile polygon editor

akien-mga committed:
Merge pull request #70868 from Maran23/4-x-scrollcontainer-custom-step-editor

Expose horizontal/vertical `custom_step` as editor property for the `ScrollContainer`

akien-mga committed:
Merge pull request #74264 from timothyqiu/self-awareness

Allow EditorInspector to change its property name style when necessary

akien-mga committed:
Merge pull request #74945 from ChibiDenDen/patch-3

Fix gles3 rendering on android studio emulator

akien-mga committed:
Merge pull request #75937 from RandomShaper/threaded_render_load

Allow creation of rendering buffers at any time

akien-mga committed:
Merge pull request #76379 from Black-Cat/fix-capsule2d-sleep

Fix precision in physics supports generation

akien-mga committed:
Merge pull request #76418 from reduz/method-bind-validated-call

Add ValidatedCall to MethodBind

akien-mga committed:
Merge pull request #76536 from AThousandShips/bit_poly_size

Fix size error in `BitMap.opaque_to_polygons`

akien-mga committed:
Merge pull request #76540 from reduz/redo-remote-filesystem

Redo how the remote filesystem works

akien-mga committed:
CI: Use gh-cli for changed files, and workaround codespell skip list bug

For PRs, this should give a more accurate list, as the previous method would
diff to the tip of the `master` branch, which could include new commits (and
thus changed files) not present in the PR branch.

codespell's `--skip` option doesn't work at all with folders when used
together with an explicit list of paths to work with, so let's not use it.

akien-mga committed:
Merge pull request #57179 from spacechase0/notify-world2d-changed-master

Propagate previously unused NOTIFICATION_WORLD_2D_CHANGED, make CanvasItem/CollisionObject2D use it

akien-mga committed:
Merge pull request #64120 from kleonc/local-vector-tight

Make `LocalVector` respect its `tight` template parameter

akien-mga committed:
Merge pull request #75751 from KoBeWi/eradicate_ViewportTexture_errors_entirely_and_surely

Prevent errors when using ViewportTexture

akien-mga committed:
Merge pull request #76675 from rburing/csg_debug_collision

Add debug collision shape to CSG with collision

akien-mga committed:
Merge pull request #76842 from akien-mga/ci-codespell-fix-skip-list

CI: Use gh-cli for changed files, and workaround codespell skip list bug

HolonProduction committed:
Don't expose mac specific display server.

jmb462 committed:
Fix regression with right click on main selection

akien-mga committed:
Merge pull request #76865 from jmb462/multicarets_selection_popup

Fix regression with right click on main selection

chfoo committed:
Ensure WindowData minimized/maximized are mutually exclusive

The window manager can break the assumption that
fullscreen/maximized/minimized values are mutually exclusive.

AThousandShips committed:
Clarify that event handling does not affect `Input`

akien-mga committed:
Merge pull request #57894 from Sauermann/fix-subviewport-1

Fix unrestricted mouse-event propagation to SubViewports for Physics-Picking

akien-mga committed:
Merge pull request #70286 from the-sink/popupmenu_icon_modulate

Add icon_modulate set/get functionality to PopupMenu

akien-mga committed:
Merge pull request #70564 from stmSi/cant-set-srollcontainer-scrollvalue-in-ready

Fix: Can't set ScrollContainer `scroll_*` value in _ready()

akien-mga committed:
Merge pull request #76092 from ChibiDenDen/shader_cache_gles3

Add shader cache to gles3

akien-mga committed:
Merge pull request #76624 from myaaaaaaaaa/movie-stat

Print CPU and GPU usage statistics after Movie Maker is finished

akien-mga committed:
Merge pull request #76791 from OmarShehata/fix-android-double-tap

Fix double tap & drag on Android

akien-mga committed:
Merge pull request #76832 from RandomShaper/cluster_render_prevail

Save cluster render shader from being optimized out entirely

lyuma committed:
gltf: Permit sparse accessors without a bufferView.

lawnjelly committed:
Make acos and asin safe

A common bug with using acos and asin is that input outside -1 to 1 range will result in Nan output. This can occur due to floating point error in the input.

The standard solution is to provide safe_acos function with clamped input. For Godot it may make more sense to make the standard functions safe.

RedworkDE committed:
Fix various errors when running the unit tests

Co-authored-by: bruvzg <7645683+bruvzg@users.noreply.github.com>

akien-mga committed:
Merge pull request #76735 from AThousandShips/natural_cmp

Add `naturalcasecmp_to` function to `String`

akien-mga committed:
Merge pull request #76755 from RandomShaper/fix_allocator_thsafe

Add missing thread safety to PagedAllocator

akien-mga committed:
Merge pull request #76814 from KoBeWi/underdata

Don't refresh inspector when changing internal meta

akien-mga committed:
Merge pull request #76868 from chfoo/issue-72728

Ensure WindowData minimized/maximized are mutually exclusive

akien-mga committed:
Merge pull request #76874 from AThousandShips/input_doc

Clarify that event handling does not affect `Input`

RandomShaper committed:
Avoid sync issues in materials with scheduled shader updates

akien-mga committed:
Merge pull request #73588 from smosages/resolve-display-shader-settings-in-settings-editor

Define shader language project settings before creation of `TextShaderEditor` object.

akien-mga committed:
Merge pull request #74711 from BastiaanOlij/add_texture_native_handle

Provide access to internal graphics handles for textures

akien-mga committed:
Merge pull request #76490 from dsnopek/dump-gdscript-docs

Dump API docs from inline GDScript comments using --doctool --gdscript-docs PATH

akien-mga committed:
Merge pull request #76661 from bonjorno7/hsl

Add API for HSL conversion

akien-mga committed:
Merge pull request #76730 from AThousandShips/doc_order

Make documentation sorting use natural order

akien-mga committed:
Merge pull request #76790 from achernik/master

LSP: don't send empty completion command

akien-mga committed:
Merge pull request #76824 from Calinou/doc-export-pck-binary-resource-conversion

Document caveats related to exported project binary resource conversion

akien-mga committed:
Merge pull request #76875 from lyuma/fix_zero_sparse_accessors

gltf: Permit sparse accessors without a bufferView.

akien-mga committed:
Merge pull request #76878 from RedworkDE/tests-fix-errors

Fix various errors when running the unit tests

AThousandShips committed:
Add missing shader debug documentation

akien-mga committed:
Merge pull request #76885 from AThousandShips/doc_update

Add missing shader debug documentation

Faless committed:
Merge pull request #75721 from ScorpionInc/Expose_String_functions_for_X509Certificate

Exposes String functions for X509Certificates

akien-mga committed:
Merge pull request #76088 from rburing/contact_points_global_3d

PhysicsDirectBodyState3D: report contact points using global coordinates

akien-mga committed:
CI: Skip codespell check if `CHANGED_FILES` is empty

This can happen when only thirdparty files are changed, since we grep them out.
Re-add `bin` and `thirdparty` to the skip list for good measure.

RandomShaper committed:
Avoid sync issues in resources with deferred updates

akien-mga committed:
Merge pull request #64388 from Rindbee/improve-ViewportTexture

Improve the UX of ViewportTexture in the editor

akien-mga committed:
Merge pull request #69988 from smix8/navigation_rvo_rework_4.x

Rework Navigation Avoidance

akien-mga committed:
Merge pull request #76532 from dalexeev/tree-fix-multiline-drawing

Fix multiline items drawing in `Tree`

akien-mga committed:
Merge pull request #76830 from ajreckof/fix_wrapf_CI_with_too_big_float

fix a test for wrapf with signed 32-bit float

akien-mga committed:
Merge pull request #76903 from akien-mga/ci-codesp-hell

CI: Skip codespell check if `CHANGED_FILES` is empty

RandomShaper committed:
Use WorkerThreadPool for threaded resource loading

and-rad committed:
Store sensitive export options in dedicated credentials file

AThousandShips committed:
Make `EditorPropertyNameProcessor` check `EditorSettings` validity

akien-mga committed:
Merge pull request #62378 from trollodel/gsoc_2022_multiwindow

Add multi window code and shader editors (GSOC'22 Project)

akien-mga committed:
Merge pull request #75901 from reduz/refactor-node-processing

Refactor Node Processing to allow Scene Multithreading

akien-mga committed:
Merge pull request #76165 from and-rad/safe-credentials

Store sensitive export options in dedicated credentials file

akien-mga committed:
Merge pull request #76905 from Faless/mbedtls/no_weak_crypto_no_patch

mbedTLS: disable weak crypto and TLS versions.

akien-mga committed:
Merge pull request #64628 from KoBeWi/a_new_meta

Add metadata to resource previews

akien-mga committed:
Merge pull request #76910 from AThousandShips/inspector_fix

Make `EditorPropertyNameProcessor` check `EditorSettings` validity

akien-mga committed:
rvo2: Include cstdint for uint32_t

Fixes #76907.

akien-mga committed:
Merge pull request #76913 from akien-mga/rvo2-include-cstdint

rvo2: Include cstdint for uint32_t

stolk committed:
x11: Do not use uninitialized timestamps.

Fixes #76922

Sauermann committed:
Propagate shortcuts to SubViewports

`push_unhandled_input` no longer propagates events to `SubViewports`.

Geometror committed:
Improve editor state persistence

akien-mga committed:
Merge pull request #74405 from RandomShaper/fix_res_loader

Fix multi-threaded resource loading

akien-mga committed:
Merge pull request #72277 from Geometror/improve-editor-state-preservation

Improve editor state persistence

akien-mga committed:
Merge pull request #76592 from spanzeri/fix-skeletons-and-bones

Fix skeleton_3d & physical_bone_3d editor errors

akien-mga committed:
Merge pull request #76794 from Wiwip/inline-edit

Inline editor for the file system dock

akien-mga committed:
Merge pull request #76812 from brno32/curve-3d-tests

Add unit tests for Curve3D

akien-mga committed:
Merge pull request #76826 from AThousandShips/item_list_fix

Minor bugfixes to `ItemList`

akien-mga committed:
Merge pull request #76856 from HolonProduction/macos-specific

Don't expose macOS and iOS specific display servers.

akien-mga committed:
Merge pull request #76924 from stolk/fixuninit

x11: Do not use uninitialized timestamps.

akien-mga committed:
Merge pull request #76926 from Sauermann/fix-call-to-unhandled-input

Propagate shortcut events to SubViewports

akien-mga committed:
Merge pull request #75678 from smix8/navregion_update_guards_4.x

Prevent unnecessary navigation map synchronizations

akien-mga committed:
Merge pull request #76906 from lawnjelly/safe_acos4_2

Make acos and asin safe

radzo73 committed:
Merge branch 'master' of https://github.com/radzo73/godot

radzo73 committed:
Revert "Merge branch 'master' of https://github.com/radzo73/godot"

This reverts commit cb5673bf6ec64ce1f113fa2502ee30f59bf7543e, reversing
changes made to fd4a06c51555904104b18494d0224f450d74fe2a.

radzo73 commented:
System-exclusive problems that I have to deal with when encoding/decoding to Godot's `Image`:
- Windows [requires a clipboard format](https://learn.microsoft.com/en-us/windows/win32/dataxchg/standard-clipboard-formats#constants) when getting/setting data. Most likely format for image purposes is `CF_DIB`.
- MacOS' clipboard primarily uses two image formats, [PNG](https://developer.apple.com/documentation/appkit/nspasteboardtypepng) and [TIFF](https://developer.apple.com/documentation/appkit/nspasteboardtypetiff)
- iOS can easily [get all images on the clipboard](https://developer.apple.com/documentation/uikit/uipasteboard/1622086-images?language=objc), but main image formats will be `kUTTypePNG ` and `kUTTypeJPEG`.
- X11, Android, Web...I have no idea. (there's also the Wayland port to worry about)

radzo73 commented:
Closing in favor of a fresh fork
(because I didn't change those 490 files, Git)

aaronfranke commented:
@radzo73 In the future, you should open a pull request from a separate branch (`radzo73:something`), not from master (`radzo73:master`). This would prevent the issues you had.
